packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -613,36 +618,68 @@ class _FocusableActionDetectorState extends State<FocusableActionDetector> {
   void _handleMouseEnter(PointerEnterEvent event) {
     assert(widget.onShowHoverHighlight != null);
     if (!_hovering) {
-      // TODO(gspencergoog): remove scheduleMicrotask once MouseRegion event timing has changed.
-      scheduleMicrotask(() { setState(() { _hovering = true; _handleShowHoverHighlight(); }); });
+      _mayTriggerCallback(task: () {
+        setState(() { _hovering = true; });
+      });
     }
   }
 
   void _handleMouseExit(PointerExitEvent event) {
     assert(widget.onShowHoverHighlight != null);
     if (_hovering) {
-      // TODO(gspencergoog): remove scheduleMicrotask once MouseRegion event timing has changed.
-      scheduleMicrotask(() { setState(() { _hovering = false; _handleShowHoverHighlight(); }); });
+      _mayTriggerCallback(task: () {
+        setState(() { _hovering = false; });
+      });
     }
   }
 
   bool _focused = false;
   void _handleFocusChange(bool focused) {
     if (_focused != focused) {
-      setState(() {
-        _focused = focused;
-        _handleShowFocusHighlight();
-        widget.onFocusChange?.call(_focused);
+      _mayTriggerCallback(task: () {
+        setState(() {
+          _focused = focused;
+          widget.onFocusChange?.call(_focused);
+        });
       });
     }
   }
 
-  void _handleShowHoverHighlight() {
-    widget.onShowHoverHighlight?.call(_hovering && widget.enabled && _canShowHighlight);
+  bool _shouldShowHoverHighlight({@required FocusableActionDetector target}) {
+    return _hovering && target.enabled && _canShowHighlight;
+  }
+  bool _shouldShowFocusHighlight({@required FocusableActionDetector target}) {
+    return _focused && target.enabled && _canShowHighlight;
+  }
+
+  // Record old states, do `task` if not null, then compare old states with the
+  // new state, and trigger callbacks if necessary.
+  //
+  // The old states are collected on `oldWidget` if it is provided, or the
+  // current widget (before doing `task`) otherwise.
+  void _mayTriggerCallback({VoidCallback task, FocusableActionDetector oldWidget}) {
+    assert(SchedulerBinding.instance.schedulerPhase != SchedulerPhase.persistentCallbacks);
+    final FocusableActionDetector oldTarget = oldWidget ?? widget;
+    final bool didShowHoverHighlight = _shouldShowHoverHighlight(target: oldTarget);
+    final bool didShowFocusHighlight = _shouldShowFocusHighlight(target: oldTarget);
+    if (task != null)
+      task();
```

Comment:
If the task is calling `setState`, isn't it possible that it won't actually execute the contents of the `setState` function arg right away? Although, I'm not sure anything needs to call `setState` in the task anymore, so it may be moot.

Reply:
I think according to the doc of `setState` the content is run synchronously.